apps/condo/domains/common/utils/serverSchema/changeTrackable.js

```diff
@@ -0,0 +1,355 @@
+ * @param {BuildDataToStoreChangeFromArgs} args
+ */
+const buildDataToStoreChangeFrom = async (args) => {
+    const {
+        existingItem,
+        updatedItem,
+        context,
+        originalInput,
+        fields,
+        displayNameResolvers,
+        relatedManyToManyResolvers,
+    } = args
+    const data = {}
+    // Since `map` uses a series of async function calls, we need to use `Promise.all`,
+    // otherwise, final result will miss fields, calculated asynchronously.
+    // https://stackoverflow.com/questions/47065444/lodash-is-it-possible-to-use-map-with-async-functions
+    await Promise.all(Object.keys(fields).map(async (key) => {
+        const field = fields[key]
+        if (isScalar(field)) {
+            if (existingItem[key] !== updatedItem[key]) {
+                data[`${ key }From`] = existingItem[key]
+                data[`${ key }To`] = updatedItem[key]
+            }
+        } else if (isRelationSingle(field)) {
+            if (existingItem[key] !== updatedItem[key]) {
+                data[`${ key }IdFrom`] = existingItem[key]
+                data[`${ key }IdTo`] = updatedItem[key]
+                data[`${ key }DisplayNameFrom`] = await displayNameResolvers[key](existingItem[key])
+                data[`${ key }DisplayNameTo`] = await displayNameResolvers[key](updatedItem[key])
+            }
+        } else if (isRelationMany(field)) {
+            if (originalInput[key]) {
+                // Since many-to-many relation is stored in different schema, there is no
+                // direct information here about initial list of related items.
+                // As an easy solution, we can utilize `originalInput`, that have
+                // relation "Nested mutations", like `connect` and `disconnect`
+                // https://www.keystonejs.com/keystonejs/fields/src/types/relationship/#nested-mutations
+                const { existing, updated } = await relatedManyToManyResolvers[key]({
+                    context,
+                    existingItem,
+                    originalInput,
+                })
+                if (_.difference(existing.ids, updated.ids).length > 0) {
+                    data[`${ key }IdsFrom`] = existing.ids
+                    data[`${ key }IdsTo`] = updated.ids
+                    data[`${ key }DisplayNamesFrom`] = existing.displayNames
+                    data[`${ key }DisplayNamesTo`] = updated.displayNames
+                }
+
+            }
+        }
+    }))
+    return data
+}
+
+const isScalar = (field) => (
+    field.type !== Relationship
+)
+
+const isRelationSingle = (field) => (
+    field.type === Relationship && !field.many
+)
+
+const isRelationMany = (field) => (
+    field.type === Relationship && field.many
+)
+
+const mapScalars = (acc, value, key) => {
+    acc[`${key}From`] = mapScalar(value)
+    acc[`${key}To`] = mapScalar(value)
+}
+
+const mapScalar = (field) => (
+    _.pick(field, ['schemaDoc', 'type'])
```

Comment:
what about `options` ? for Select type
And what about specific type arguments?

Reply:
It seems, like it's not needed, because we just store a raw value.
It's not for a form in UI.